tools/Makefile.unix

```diff
@@ -565,17 +547,15 @@ do_gconfig: dirlinks apps_preconfig
 
 gconfig: do_gconfig clean_context
 
-do_savedefconfig: dirlinks apps_preconfig
-
-savedefconfig: do_savedefconfig
+savedefconfig: dirlinks apps_preconfig
 	$(Q) APPSDIR=${CONFIG_APPS_DIR} kconfig-conf --savedefconfig defconfig.tmp Kconfig
 	$(Q) sed -i -e "/CONFIG_APPS_DIR=/d" defconfig.tmp
 	$(Q) grep "CONFIG_ARCH=" .config >> defconfig.tmp
 	$(Q) grep "^CONFIG_ARCH_CHIP_" .config >> defconfig.tmp; true
-	$(Q) grep "^CONFIG_ARCH_CHIP=" .config >> defconfig.tmp; true
-	$(Q) grep "CONFIG_ARCH_BOARD=" .config && grep "CONFIG_ARCH_BOARD=" .config >> defconfig.tmp; true
-	$(Q) grep "^CONFIG_ARCH_CUSTOM" .config && grep "^CONFIG_ARCH_CUSTOM" .config >> defconfig.tmp; true
-	$(Q) grep "^CONFIG_ARCH_BOARD_CUSTOM=" .config && grep "^CONFIG_ARCH_BOARD_CUSTOM" .config >> defconfig.tmp; true
+	$(Q) grep "CONFIG_ARCH_CHIP=" .config >> defconfig.tmp; true
+	$(Q) grep "CONFIG_ARCH_BOARD=" .config >> defconfig.tmp; true
+	$(Q) grep "^CONFIG_ARCH_CUSTOM" .config >> defconfig.tmp; true
+	$(Q) grep "^CONFIG_ARCH_BOARD_CUSTOM" .config >> defconfig.tmp; true
```

Comment:
@xiaoxiang781216  - this logic has been broken several times in the past.  Did you test these changes on a custom board with the complete Kconfig life cycle oldconfig, menuconfig, savedefconfig?   

Is CONFIG_ARCH_CUSTOM used?

```
CONFIG_ARCH_BOARD_CUSTOM=y
CONFIG_ARCH_BOARD_CUSTOM_DIR="../nuttx-config"
CONFIG_ARCH_BOARD_CUSTOM_DIR_RELPATH=y
CONFIG_ARCH_BOARD_CUSTOM_NAME="tvc15-pro"
CONFIG_ARCH_CHIP="stm32h7"
CONFIG_ARCH_CHIP_STM32H743ZI=y
CONFIG_ARCH_CHIP_STM32H7=y
```

Reply:
> @xiaoxiang781216 - this logic has been broken several times in the past. Did you test these changes on a custom board with the complete Kconfig life cycle oldconfig, menuconfig, savedefconfig?
> 

No, I don't have custom board environment. I just test savedefconfig in our internal project.

> Is CONFIG_ARCH_CUSTOM used?
>

I don't know  I just keep it as before except remove the && check.

> ```
> CONFIG_ARCH_BOARD_CUSTOM=y
> CONFIG_ARCH_BOARD_CUSTOM_DIR="../nuttx-config"
> CONFIG_ARCH_BOARD_CUSTOM_DIR_RELPATH=y
> CONFIG_ARCH_BOARD_CUSTOM_NAME="tvc15-pro"
> CONFIG_ARCH_CHIP="stm32h7"
> CONFIG_ARCH_CHIP_STM32H743ZI=y
> CONFIG_ARCH_CHIP_STM32H7=y
> ```

Does PX4 use custom board? If so, it's better to add a test project use this feature, otherwise it is hard to ensure that we never break this feature without some test.

Reply:
Yes.  Let me test this....

Reply:
On master and this branch the version.sh fails out of a git tree.  I have lost track of it's requirements and intent.

Here is the hack I used to get past it. 

```
diff --git a/tools/version.sh b/tools/version.sh
index a53b3cfbcc..5988186f2c 100755
--- a/tools/version.sh
+++ b/tools/version.sh
@@ -85,8 +85,15 @@ done
 
 OUTFILE=$1
 
+INGIT=$(git rev-parse --is-inside-work-tree)
+
 if [ -z ${VERSION} ] ; then
-  VERSION=`git tag --sort=taggerdate | tail -1 | cut -d'-' -f2`
+  if [ "X${INGIT}" = "Xtrue" ] ; then
+    VERSION=`git tag --sort=taggerdate | tail -1 | cut -d'-' -f2`
+  else
+    VERSION="0.0"
+    BUILD="na"
+  fi
 fi
 
 # Make sure we know what is going on
```

This PR breaks a defconfig on make oldconfig

![image](https://user-images.githubusercontent.com/1945821/77163814-26e86c80-6a6c-11ea-88eb-d6ad969a0cd6.png)

Reply:
> On master and this branch the version.sh fails out of a git tree. I have lost track of it's requirements and intent.
> 
> Here is the hack I used to get past it.
> 
> ```
> diff --git a/tools/version.sh b/tools/version.sh
> index a53b3cfbcc..5988186f2c 100755
> --- a/tools/version.sh
> +++ b/tools/version.sh
> @@ -85,8 +85,15 @@ done
>  
>  OUTFILE=$1
>  
> +INGIT=$(git rev-parse --is-inside-work-tree)
> +
>  if [ -z ${VERSION} ] ; then
> -  VERSION=`git tag --sort=taggerdate | tail -1 | cut -d'-' -f2`
> +  if [ "X${INGIT}" = "Xtrue" ] ; then
> +    VERSION=`git tag --sort=taggerdate | tail -1 | cut -d'-' -f2`
> +  else
> +    VERSION="0.0"
> +    BUILD="na"
> +  fi
>  fi
>  
>  # Make sure we know what is going on
> ```
> 

The out of tree build is fine in my environment.

> This PR breaks a defconfig on make oldconfig
> 
> ![image](https://user-images.githubusercontent.com/1945821/77163814-26e86c80-6a6c-11ea-88eb-d6ad969a0cd6.png)

Please try the new PR, this problem should be fixed.

Reply:
@xiaoxiang781216 

I am assuming you want me to re-test this PR now that you force pushed or is there a new one?

Reply:
> The out of tree build is fine in my environment.

It is not out of tree, it is not under git. 

Does the tools still work without the .git dir?  It use to.

Reply:
> > The out of tree build is fine in my environment.
> 
> It is not out of tree, it is not under git.
> 
> Does the tools still work without the .git dir? It use to.

Sorry, make you confuse, my "out of tree" mean no git. Here is my test step:
```
mv .git .git.bak
./tools/configure.sh sim:nsh
```
The result is same without mv

Reply:
Oh OK.

hmm - good to know you have it working - when we do an uptake I will figure out the cause and fix it. 

The other bug was a real problem because it breaks every defconfig on an uptake. Thank you for fixing it!